src/cart/crt_hg.c

```diff
@@ -42,6 +42,146 @@
 
 #include "crt_internal.h"
 
+#ifdef HASH_TEST
+int hg_addr_table_index=0;
+#endif
+
+struct ha_entry {
+        hg_class_t      *hg_class;
```

Comment:
hg_class is associated with crt_context. As such we should keep the addr table per context and not store hg_class for each entry. In your current implementation (even if client id was unique), you would run into issues, since if client sends to server on address rank=0 tag=0 and on rank=0 tag=1, you will end up with 2 ha_entries for the same client id.

Storing this hash table per each context will save space.

Whenever hg_class is needed it can be derived from context in question from crt_ctx->cc_hg_ctx.chc.chc_hgcla